README.md
# jrvFinance

This R package implements the basic financial analysis functions
similar to (but not identical to) what is available in most
spreadsheet software. This includes finding the IRR and NPV of
regularly spaced cash flows and annuities. Bond pricing and YTM
calculations are included. In addition, Black Scholes option pricing
and Greeks are also provided.

## NPV, XNPV, IRR and XIRR functions

```{r}
npv(cf=c(100,250,300), rate=5e-2)
npv(cf=c(1,3,2), rate=10e-2, cf.t=c(0.3,1.9,2.5))
irr(c(-600,300,400))
irr(cf=c(-450,100,300,200), cf.t=c(0, 0.3,1.9,2.5)) 
```

## Annuity functions

```{r}
annuity.pv(rate=10e-2, n.periods=15)
annuity.pv(rate=10e-2, n.periods=15, immediate.start = TRUE)
annuity.pv(rate=10e-2, instalment = 450, n.periods=360, cf.freq=12, comp.freq=2)
annuity.rate(pv=50000, instalment = 450, n.periods=360, cf.freq=12, comp.freq=2)
annuity.instalment(rate=9e-2, pv=10000, n.periods=8)
annuity.instalment.breakup(rate=9e-2, pv=10000, n.periods=8, period.no=5)
```

## Bond Price, Yield and Duration

```{r}
bond.price(settle="2012-04-15", mature="2022-01-01", coupon=8e-2,
           yield=8.8843e-2)
bond.price(settle="2012-04-15", mature="2022-01-01", coupon=8e-2,
bond.price(settle="2012-04-15", mature="2022-01-01", coupon=8e-2,
           yield=8.8843e-2, freq=1, comp.freq=2)
bond.yield(settle="2012-04-15", mature="2022-01-01", coupon=8e-2,
           price=95) 
bond.duration(settle="2012-04-15", mature="2022-01-01", coupon=8e-2,
              yield=8.8843e-2)
bond.duration(settle="2012-04-15", mature="2022-01-01", coupon=8e-2,
              yield=8.8843e-2, modified=TRUE)
coupons.dates(settle="2012-04-15", mature="2022-01-01")
coupons.next(settle="2012-04-15", mature="2022-04-01")
coupons.prev(settle="2012-04-15", mature="2022-04-01")
coupons.n(settle="2012-04-15", mature="2017-07-01")
```

##  (Generalized) Black Scholes Formulas


```{r}
GenBS(s=100, X=100, r=0.1, Sigma=20e-2, t=1, div_yield=0)
GenBS(s=100, X=120, r=0.1, Sigma=15e-2, t=1, div_yield=5.8e-2)
GenBSImplied(s=100, X=900, r=0, price=7.97, t=1, div_yield=0)
```

## Utility functions


```{r}
equiv.rate(10e-2, from.freq = 12, to.freq = 2) 
equiv.rate(15e-2, from.freq = 1, to.freq = Inf)
edate("2005-05-17", -8) 
edate("2007-02-28", 4) 
```

## Newton Raphson and bisection root solver

The package implements a Newton Raphson root solver that is used
internally to calculate IRR and YTM. It is available for general use.


```{r} 
fn1 <-function(x){list(value=sin(x)-cos(x), gradient=cos(x)+sin(x))} 
newton.raphson.root(fn1) 
```

The package implements a bisection root solver that does a geometric
grid search to bracket the root and then calls `uniroot` to find the
root within this interval. The package uses the function internally to
calculate IRR and YTM, but `bisection.root` is available for general
use.

```{r}
bisection.root(sin, guess = 7, lower=1, upper=13)
bisection.root(sin, guess = 12, lower=1, upper=13) 
```



NEWS.md
# jrvFinance 1.4.3

Fixed bug: [immediate.start=TRUE has no effect on annuity.installment when given fv](https://github.com/jrvarma/jrvFinance/issues/2)

# jrvFinance 1.4.2

Dependency fixed. Suggests `markdown` and `rmarkdown` as required by <https://github.com/yihui/knitr/issues/1864>

# jrvFinance 1.4.1

Bug fix release. The `edate` function was giving wrong answers under certain conditions when called with vector inputs. The incorrect `if` statement has now been changed to the correct `ifelse`. Thanks to Thethach Chuaprapaisilp for finding the bug.

# jrvFinance 1.4.0

* All the bond functions now accept an additional parameter `redemption_value` to handle the situation where the bond is expected to be called at premium to par. This is based on code provided by Tom McGinty. (#1)


R/dcf.R
##' Net Present Value
##' 
##' Computes NPV (Net Present Value) for cash flows with different cash flow and
##' compounding conventions. Cash flows need not be evenly spaced.
##' 
##'
##' @param cf Vector of cash flows
##' @param cf.t Optional vector of timing (in years) of cash flows. If omitted
##' regular sequence of years is assumed.
##' @param rate The interest rate in decimal (0.10 or 10e-2 for 10\%)
##' @param cf.freq Frequency of annuity payments: 1 for annual, 2 for
##' semi-annual, 12 for monthly.
##' @param comp.freq Frequency of compounding of interest rates: 1 for annual,
##' 2 for semi-annual, 12 for monthly, Inf for continuous compounding.
##' @param immediate.start Logical variable which is \code{TRUE} when the first cash
##' flows is at the beginning of the first period (for example, immediate
##' annuities) and \code{FALSE} when the first cash flows is at the end of the first
##' period (for example, deferred annuities)
##' @export
npv <- function(cf, rate, cf.freq=1, comp.freq=1,
                cf.t = seq(from = if (immediate.start) 0 else 1/cf.freq,
                  by = 1/cf.freq, along.with = cf),
                immediate.start=FALSE){
  cc.rate <- equiv.rate(rate, from.freq=comp.freq, to.freq=Inf)
  df <- exp(-cc.rate * cf.t)
  as.numeric(crossprod(cf, df))
}


##' Internal Rate of Return
##' 
##' Computes IRR (Internal Rate of Return) for cash flows with different cash flow and
##' compounding conventions. Cash flows need not be evenly spaced.
##' 
##'
##' @inheritParams npv
##' @param interval the interval c(lower, upper) within which to
##'     search for the IRR
##' @param r.guess the starting value (guess) from which the solver
##'     starts searching for the IRR
##' @param toler the argument \code{toler} for
##'     \code{\link{irr.solve}}.  The IRR is regarded as correct if
##'     abs(NPV) is less than \code{toler}.  Otherwise the \code{irr}
##'     function returns \code{NA}
##' @param convergence the argument \code{convergence} for
##'     \code{\link{irr.solve}}
##' @param max.iter the argument \code{max.iter} for
##'     \code{\link{irr.solve}}
##' @param method The root finding method to be used. The
##'     \code{default} is to try Newton-Raphson method
##'     (\code{\link{newton.raphson.root}}) and if that fails to try
##'     bisection (\code{\link{bisection.root}}). The other two
##'     choices (\code{newton} and \code{bisection} force only one of
##'     the methods to be tried.
##' @export
irr <- function(cf, interval = NULL, cf.freq = 1, comp.freq = 1,
                cf.t = seq(from = 0, by = 1/cf.freq, along.with = cf),
                r.guess = NULL, toler = 1e-6, convergence = 1e-8,
                max.iter = 100,
                method = c('default', 'newton', 'bisection')){

    if (sign(max(cf)) == sign(min(cf)))
        return(NA)
    fn <- function(r){
        df <- exp(-r*cf.t)
        list(value=as.numeric(crossprod(cf, df)),
             gradient=-as.numeric(crossprod(cf*df,cf.t)))
    }
    method <- match.arg(method)
    res <- NA
    tryCatch(
        res <- irr.solve(f=fn, interval=interval, r.guess=r.guess, 
                         toler = toler, convergence = convergence,
                         max.iter = max.iter, method = method),
        warning = function(war){}
    )
    if(is.na(res)){
        warning(.irr.warning.msg(method))
    }
    equiv.rate(res, from.freq=Inf, to.freq=comp.freq)
}


##' Equivalent Rates under different Compounding Conventions
##' 
##' Converts an interest rate from one compounding convention to
##' another (for example from semi-annual to monthly compounding or
##' from annual to continuous compounding)
##' 
##'
##' @inheritParams npv
##' @param from.freq Frequency of compounding of the given interest
##'     rate: 1 for annual, 2 for semi-annual, 12 for monthly, Inf for
##'     continuous compounding.
##' @param to.freq Frequency of compounding to which the given
##'     interest rate is to be converted: 1 for annual, 2 for
##'     semi-annual, 12 for monthly, \code{Inf} for continuous
##'     compounding.
##' @export
equiv.rate <- function(rate, from.freq = 1, to.freq = 1){
  cc.rate <- ifelse(from.freq == Inf, rate,
                    log(1+rate/from.freq)*from.freq)
  if (to.freq == Inf) cc.rate else (exp(cc.rate/to.freq)-1)*to.freq
}

##' Duration and Modified Duration
##' 
##' Computes Duration and Modified Duration for cash flows with
##' different cash flow and compounding conventions. Cash flows need
##' not be evenly spaced.
##' 
##'
##' @inheritParams npv
##' @param modified in function duration, \code{TRUE} if modified
##'     duration is desired. \code{FALSE} otherwise.
##' @export
duration <- function(cf, rate, cf.freq=1, comp.freq=1,
                     cf.t = seq(from = ifelse(immediate.start, 0,
                                              1/cf.freq),
                       by = 1/cf.freq, along.with = cf),
                     immediate.start=FALSE, modified=FALSE){
  cc.rate <- equiv.rate(rate, from.freq=comp.freq, to.freq=Inf)
  df <- exp(-cc.rate * cf.t)
  D <- as.numeric(crossprod(cf*df, cf.t)/crossprod(cf, df))
  D / if (modified)  1 + rate/comp.freq else 1
}



R/GenBS.R
##' Generalized Black Scholes model for pricing vanilla European options
##'
##' Compute values of call and put options as well as the Greeks -
##' the sensitivities of the option price to various input arguments using the
##' Generalized Black Scholes model. "Generalized" means that the asset can
##' have a continuous dividend yield.
##' 
##' The Generalized Black Scholes formula for call options is \cr
##' \eqn{e^{-r t} (s \;  e^{g t} \; Nd1 - X \; Nd2)}{exp(-r * t) * (s * exp(g * t) * Nd1 - X * Nd2)} \cr
##' where \cr
##' \eqn{g = r - div\_yield} \cr
##' \eqn{Nd1 = N(d1)} and \eqn{Nd2 = N(d2)} \cr
##' \eqn{d1 = \frac{log(s / X) + (g + Sigma^2/ 2)  t}{Sigma \sqrt{t}}}{d1 = (log(s / X) + (g + Sigma^2/ 2) * t) / (Sigma * sqrt(t))} \cr
##' \eqn{d2 = d1 - Sigma \sqrt{t}}{d2 = d1 - Sigma * sqrt(t)} \cr
##' N denotes the normal CDF (\code{\link{pnorm}})\cr
##' For put options, the formula is \cr
##' \eqn{e^{-r t}  (-s \; e^{g  t} \; Nminusd1 + X \;  Nminusd2)}{exp(-r * t) * (-s * exp(g * t) * Nminusd1 + X * Nminusd2)}\cr
##' where \cr
##' \eqn{Nminusd1 = N(-d1)} and \eqn{Nminusd2 = N(-d2)} \cr
##'
##' @param s the spot price of the asset (the stock price for options on stocks)
##' @param X the exercise or strike price of the option
##' @param r the continuously compounded rate of interest in decimal (0.10 or 10e-2 for 10\%)
##' (use \code{\link{equiv.rate}} to convert to a continuously compounded rate)
##' @param Sigma the volatility of the asset price in decimal  (0.20 or 20e-2 for 20\%)
##' @param t the maturity of the option in years
##' @param div_yield the continuously compounded dividend yield (0.05 or 5e-2 for 5\%)
##' (use \code{\link{equiv.rate}} to convert to a continuously compounded rate)
##' @return A list of the following elements
##' \item{call}{the value of a call option}
##' \item{put}{the value of a put option}
##' \item{Greeks}{a list of the following elements}
##' \item{Greeks$callDelta}{the delta of a call option - the sensitivity to the spot price of the asset}
##' \item{Greeks$putDelta}{the delta of a put option - the sensitivity to the spot price of the asset}
##' \item{Greeks$callTheta}{the theta of a call option - the time decay of the option value
##' with passage of time. Note that time is measured in years. To find a daily theta divided by 365.}
##' \item{Greeks$putTheta}{the theta of a put option}
##' \item{Greeks$Gamma}{the gamma of a call or put option - the second derivative with respect to the spot price
##' or the sensitivity of delta to the spot price}
##' \item{Greeks$Vega}{the vega of a call or put option - the sensitivity to the volatility}
##' \item{Greeks$callRho}{the rho of a call option - the sensitivity to the interest rate}
##' \item{Greeks$putRho}{the rho of a put option - the sensitivity to the interest rate}
##' \item{extra}{a list of the following elements}
##' \item{extra$d1}{the d1 of the Generalized Black Scholes formula}
##' \item{extra$d2}{the d2 of the Generalized Black Scholes formula}
##' \item{extra$Nd1}{is \code{\link{pnorm}}(d1)}
##' \item{extra$Nd2}{is \code{\link{pnorm}}(d2)}
##' \item{extra$Nminusd1}{is \code{\link{pnorm}}(-d1)}
##' \item{extra$Nminusd2}{is \code{\link{pnorm}}(-d2)}
##' \item{extra$callProb}{the (risk neutral) probability that the call will be exercised = \code{Nd2}}
##' \item{extra$putProb}{the (risk neutral) probability that the put will be exercised = \code{Nminusd2}}
##'
##' @export
##' @importFrom stats pnorm dnorm
GenBS <- function(s, X, r, Sigma, t, div_yield = 0){
  g = r - div_yield ## the growth rate of the asset price
  d1 <- (log(s / X) + (g + Sigma*Sigma / 2) * t) / (Sigma * sqrt(t))
  ## the division above can be 0/0=NaN if s=X and either t=0 or g=Sigma=0
  ## we set d1 to 0 in this case since numerator is of higher order in t and Sigma
  d1 <- ifelse(is.nan(d1), 0, d1)
  d2 <- d1 - Sigma * sqrt(t)
  Nd1 <- pnorm(d1)
  Nd2 <- pnorm(d2)
  Nminusd1 <- pnorm(-d1)
  Nminusd2 <- pnorm(-d2)
  ## Black Scholes call price
  call <- exp(-r * t) * (s * exp(g * t) * Nd1 - X * Nd2)
  put <- exp(-r * t) * (-s * exp(g * t) * Nminusd1 + X * Nminusd2)
  callDelta <- exp(-div_yield*t)*Nd1
  putDelta <- -exp(-div_yield*t)*Nminusd1
  a <- -s * dnorm(d1) * Sigma * exp(-div_yield*t) / (2 * sqrt(t))
  ## if t is 0, the above is 0/0 = NaN but dnorm(d1) term goes to 0 faster
  ## so we set the result to 0 in this case
  a <- ifelse(is.nan(a), 0,  a)
  b = r * X * exp(-r * t) * Nd2
  c = div_yield * s * Nd1* exp(-div_yield*t)
  callTheta <- a - b + c
  b = r * X * exp(-r * t) * Nminusd2
  c = div_yield * s * Nminusd1 * exp(-div_yield*t)
  putTheta <- a + b - c
  Gamma <- dnorm(d1) * exp(-div_yield*t)/ (s * Sigma * sqrt(t))
  ## if t is 0, the above is 0/0 = NaN but dnorm(d1) term goes to 0 faster
  ## so we set the result to 0 in this case
  Gamma <- ifelse(is.nan(Gamma), 0,  Gamma)
  Vega <- s * sqrt(t) * dnorm(d1) * exp(-div_yield*t)
  callRho <- X * t * exp(-r * t) * Nd2
  putRho <- -X * t * exp(-r * t) * Nminusd2
  callProb <- Nd2
  putProb <- Nminusd2
  Greeks <- list(callDelta=callDelta, putDelta=putDelta, callTheta=callTheta,
                 putTheta=putTheta, Gamma=Gamma, Vega=Vega, callRho=callRho, putRho=putRho)
  extra <- list(d1=d1, d2=d2, Nd1=Nd1, Nd2=Nd2, Nminusd1=Nminusd1,
                Nminusd2=Nminusd2, callProb=callProb, putProb=putProb)
  list(call=call, put=put, Greeks=Greeks, extra=extra)
}

##' Generalized Black Scholes model implied volatility
##'
##' Find implied volatility given the option price using the generalized Black Scholes model.
##' "Generalized" means that the asset can have a continuous dividend yield.
##' 
##' \code{GenBSImplied} calls \code{\link{newton.raphson.root}} and
##' if that fails \code{\link{uniroot}}
##'
##' @param price the price of the option
##' @param PutOpt \code{TRUE} for put options, \code{FALSE} for call options
##' @param toler passed on to \code{\link{newton.raphson.root}}
##' The implied volatility is regarded as correct if the solver is able to
##' match the option price to within less than \code{toler}. Otherwise the function returns \code{NA}
##' @param max.iter passed on to \code{\link{newton.raphson.root}}
##' @param convergence passed on to \code{\link{newton.raphson.root}}
##' @inheritParams GenBS
##'
##' @export
##' @importFrom stats uniroot
GenBSImplied <- function(s, X, r, price, t, div_yield, PutOpt=FALSE,
                         toler=1e-6, max.iter=100, convergence=1e-8){
  ## discount the exercise price to eliminate r
  X = X * exp(-r * t)
  ## adjust the spot price to eliminate div_yield
  s = s * exp(-div_yield * t)
  ## use put call parity to convert put option into call option
  if (PutOpt) price = price + (s - X)
  SminusX = s - X
  SplusX = s + X
  if (price < SminusX || price < 0 || price > s){
    warning("Implied volatility is undefined because price is outside theoretical bounds")
    return(NA)
  }
  if (price == SminusX || price == 0) 
    ## if price equals intrinsic value, volatility is zero
    return(0)
  if (X == 0 && price != s) {
    ## if x is 0, option price must equal stock price
    warning("Implied volatility is undefined because price is outside theoretical bounds")
    return(NA)
  }
  ## we use a Taylor series approximation for an initial guess
  guess <- GenBSImpliedGuess(price, SminusX, SplusX, t)
  ## since price exceeds intrinsic value, 0 is a lower bound for the volatility
  ## we now seek an upper bound by doubling 100% volatility until the price is exceeded
  upper <- 100e-2
  while (GenBS(s, X, 0, upper, t, 0)$call < price && upper < .Machine$integer.max) upper <-  upper * 2
  if (upper > .Machine$integer.max) upper <-  Inf
  f <- function(sigma){
    temp <- GenBS(s, X, 0, sigma, t, 0)
    list(value = temp$call - price, gradient = temp$Greeks$Vega)
  }
  res <- newton.raphson.root(f=f, guess=guess, lower=0, upper=upper, max.iter=max.iter,  
                           toler=toler, convergence=convergence)
  if (! is.na(res))
    return (res)
  ## if newton.raphson.root fails, we try bisection
  res2 <- uniroot(function(x){f(x)$value}, c(0, upper), tol=convergence)
  if (abs(res2$f.root > toler)){
    warning("Error finding implied volatility")
    return (NA)                         # uniroot also failed so return NA
  }
  warning("Found implied volatility using uniroot after newton.raphson.root failed")
  ## Above warning is to prevent confusion from warning message in newton.raphson.root
  return (res2$root)
}


 GenBSImpliedGuess <- function(price, SminusX, SplusX, t){
   ## We use a Taylor series approximation similar to
   ## Brenner, H. and Subrahmanyam, M. G. (1994), "A simple approximation to option valuation
   ## and hedging in the Black Scholes model", Financial Analysts Journal, Mar-Apr 1994, 25-28
   ## Assume a call option and that r has been eliminated by discounting the exercise price 
   ## Approximate log(s/x) as 2(s-x)/(s+x) = SminusX(s-x)/H
   ## where SminusX = s-x and H=(s+x)/2.
   ## We then approximate the normal integral by a Taylor series
   ## if price is the call price, this gives the approximation
   ## price/H = SigmaRootT(1/Root2Pi +SminusX/(2*H*SigmaRootT)
   ##                            +SminusX^2/(2*H^2*SigmaRootT^2*Root2Pi)
   ## Letting z=SigmaRootT*H/Root2Pi, this yields a quadratic equation for z:
   ## z^2 - (price-SminusX/2)*z +SminusX^2/(4*Pi) = 0
   ## The solution is
   ## z = (price - SminusX/2)/2 + Radical
   ## where Radical is the square root of (price-SminusX/2)^2 - SminusX^2/Pi
   ## The linear approximation is obtained by dropping the constant term in the quadratic
   ## to give the linear equation:
   ## z = (price-SminusX/2)

  root2pi = sqrt(2 * pi)
  h = 0.5 * SplusX
  temp = (price - 0.5 * SminusX)
  radical = temp*temp - SminusX*SminusX / pi
  if (radical < 0)                   # Try Linear Approximation
    sigmaRootT = (root2pi / h) * temp
  else{                              # Try Quadratic Approximation
    radical = sqrt(radical)
    sigmaRootT = (root2pi / h) * (temp / 2 + radical)
  }
   sigmaRootT / sqrt(t)
}



R/bonds.R
##' Bond pricing using yield to maturity.
##' 
##' bond.price computes the price given the yield to maturity
##' bond.duration computes the duration given the yield to maturity
##' bond.yield computes the yield to maturity given the price
##' bond.prices, bond.durations and bond.yields are wrapper functions
##' that use \code{mapply} to vectorize bond.price, bond.duration and bond.yield
##' All arguments to bond.prices, bond.durations and bond.yields
##' can be vectors.
##' On the other hand, bond.price, bond.duration and bond.yield do not allow vectors 
##' Standard compounding and day count conventions are supported for all functions.  
##' 
##'
##' @rdname bonds
##' @name bonds
##' @aliases bond.price bond.TCF
##' bond.yield bond.duration bonds
##' bond.prices bond.durations bond.yields
##' @param settle The settlement date for which the bond is traded. Can be a
##' character string or any object that can be converted into date using
##' \code{\link{as.Date}}.
##' @param mature The maturity date of the bond. Can be a character string or
##' any object that can be converted into date using \code{\link{as.Date}}
##' @param coupon The coupon rate in decimal (0.10 or 10e-2 for 10\%)
##' @param freq The frequency of coupon payments: 1 for annual, 2 for
##' semi-annual, 12 for monthly.
##' @param comp.freq The frequency of compounding of the bond yield: 1 for
##' annual, 2 for semi-annual, 12 for monthly. Usually same as freq.
##' @param price The clean price of the bond.
##' @param redemption_value The principal amount that the bond will pay on maturity or call.
##' Typically necessary when the bond is expected to be called at premium to par.
##' @param yield The yield to maturity of the bond
##' @param convention The daycount convention
##' @param modified A logical value used in duration. \code{TRUE} to return Modified
##' Duration, \code{FALSE} otherwise
##' @return \code{bond.TCF} returns a list of three components
##' \item{t}{A vector of cash flow dates in number of years}
##' \item{cf}{A vector of cash flows}
##' \item{accrued}{The accrued interest}
##' @author Prof. Jayanth R. Varma \email{jrvarma@@iima.ac.in}

NULL


##' @rdname bonds
##' @export
bond.price <- function(settle, mature, coupon, freq=2, yield, 
                  convention=c("30/360", "ACT/ACT", "ACT/360", "30/360E"),
                  comp.freq=freq, redemption_value=100){
  settle <- as.Date(settle)
  mature <- as.Date(mature)  
  TCF <- bond.TCF(settle, mature, coupon, freq, convention, redemption_value)
  if (length(TCF$t) == 1){
      ## short government bond convention = simple interest
    return (TCF$cf[1] / (1 + yield * TCF$t[1]) - TCF$accrued)
  }
  if (comp.freq != freq){
    yield <- equiv.rate(yield, from.freq=comp.freq, to.freq=freq)
  }
  df <- (1+yield/freq)^(-freq*TCF$t)
  sum(TCF$cf*df) - TCF$accrued
}

##' @rdname bonds
##' @export
bond.yield <- function(settle, mature, coupon, freq=2, price,
                  convention=c("30/360", "ACT/ACT", "ACT/360",
                               "30/360E"),
                  comp.freq=freq, redemption_value=100){
  settle <- as.Date(settle)
  mature <- as.Date(mature)
  TCF <- bond.TCF(settle, mature, coupon, freq, convention, redemption_value)
  if (length(TCF$t) == 1){
      ## short government bond convention = simple interest
      return ( (TCF$cf[1] / (price + TCF$accrued) - 1) / TCF$t[1] )
  }
  fn <- function(r){
    df <- (1+r/freq)^(-freq*TCF$t)
    dirty <- crossprod(TCF$cf, df)
    list(value = dirty - TCF$accrued - price,
         gradient = -as.numeric(crossprod(TCF$cf*df, TCF$t))/
             (1+r/freq))
  }
  equiv.rate(irr.solve(f=fn), from.freq=freq, to.freq=comp.freq)
}

##' @rdname bonds
##' @export
bond.duration <- function(settle, mature, coupon, freq=2, yield,
                     convention=c("30/360", "ACT/ACT", "ACT/360",
                                  "30/360E"),
                     modified=FALSE, comp.freq=freq, redemption_value=100){
  settle <- as.Date(settle)
  mature <- as.Date(mature)
  TCF <- bond.TCF(settle, mature, coupon, freq, convention, redemption_value)
  if (length(TCF$t) == 1){
      ## short government bond convention = simple interest
      return(yearFraction(settle, mature))
  }
  duration(cf=TCF$cf, cf.t=TCF$t, rate=yield, cf.freq=freq,
           comp.freq=comp.freq, immediate.start=FALSE,
           modified=modified)
}

##' @rdname bonds
##' @export
bond.TCF <- function(settle, mature, coupon, freq=2,
                     convention=c("30/360", "ACT/ACT", "ACT/360", "30/360E"),
                     redemption_value=100){
  settle <- as.Date(settle)
  mature <- as.Date(mature)
  nextC <- coupons.next(settle, mature, freq)
  prevC <- coupons.prev(settle, mature, freq)
  accrued <- 100 * coupon * yearFraction(prevC, settle, prevC,
                                         nextC, freq, convention)
  start <- yearFraction(settle, nextC, prevC, nextC, freq, convention)
  t <- seq(from=start, by=1/freq, length.out=coupons.n(settle, mature,
                                                       freq))
  cf <- rep(coupon*100/freq, length(t))
  cf[length(cf)] <- redemption_value + coupon*100/freq
  list(t=t,cf=cf, accrued=accrued)
}

##' Bond pricing using yield to maturity.
##'
##' 
##' Convenience functions for finding coupon dates and number of coupons of a bond.
##' 
##'
##' @name coupons
##' @aliases coupons.n coupons.dates coupons.next coupons.prev coupons
##' @author Prof. Jayanth R. Varma \email{jrvarma@@iima.ac.in}
NULL


##' @rdname coupons
##' @inheritParams bond.price
##' @export
coupons.dates <- function(settle, mature, freq=2){
  settle <- as.Date(settle)
  mature <- as.Date(mature)
  m <- -12/freq
  n <- coupons.n(settle, mature, freq)
  edate(mature, m * (n:1 - 1))
}

##' @rdname coupons
##' @inheritParams bond.price
##' @export
coupons.n <- function(settle, mature, freq=2){
  mature <- as.Date(mature)
  settle <- as.Date(settle)
  n <- as.integer(freq * (mature - settle) / 365.25)
  m <- -12/freq
  while(edate(mature, n * m) <= settle) n <- n - 1
  while(edate(mature, (n + 1) * m) > settle) n <- n + 1
  n+1
}

##' @rdname coupons
##' @inheritParams bond.price
##' @export
coupons.next <- function(settle, mature, freq=2){
  settle <- as.Date(settle)
  mature <- as.Date(mature)
  m <- -12/freq
  n <- coupons.n(settle, mature, freq)
  edate(mature, m * (n-1))
}

##' @rdname coupons
##' @inheritParams bond.price
##' @export
coupons.prev <- function(settle, mature, freq=2){
  settle <- as.Date(settle)
  mature <- as.Date(mature)
  m <- -12/freq
  n <- coupons.n(settle, mature, freq)
  edate(mature, m * n)
}

##' @rdname bonds
##' @export
bond.prices <- function(settle, mature, coupon, freq=2, yield, 
                  convention=c("30/360", "ACT/ACT", "ACT/360", "30/360E"),
                  comp.freq=freq, redemption_value=100){
  convention <- match.arg(convention)
  mapply(bond.price, settle, mature, coupon, freq, yield, convention,
         comp.freq, redemption_value, USE.NAMES=FALSE)
}

##' @rdname bonds
##' @export
bond.yields <- function(settle, mature, coupon, freq=2, price,
                  convention=c("30/360", "ACT/ACT", "ACT/360",
                               "30/360E"),
                  comp.freq=freq, redemption_value=100){
  convention <- match.arg(convention)
  mapply(bond.yield, settle, mature, coupon, freq, price, convention,
         comp.freq, redemption_value, USE.NAMES=FALSE)
}

##' @rdname bonds
##' @export
bond.durations <- function(settle, mature, coupon, freq=2, yield,
                           convention=c("30/360", "ACT/ACT",
                                        "ACT/360", "30/360E"),
                     modified=FALSE, comp.freq=freq, redemption_value=100){
  convention <- match.arg(convention)
  mapply(bond.duration, settle, mature, coupon, freq, yield, convention,
         modified, comp.freq, redemption_value, USE.NAMES=FALSE)
}


R/jrvFinance-package.R
##' Basic Finance: NPV/IRR/annuities, bond pricing, Black Scholes 
##' 
##' 
##' This package implements the basic financial analysis functions
##' similar to (but not identical to) what is available in most
##' spreadsheet software.  This includes finding the IRR, NPV and
##' duration of possibly irregularly spaced cash flows and
##' annuities. Bond pricing, YTM and duration calculations are
##' included. Black Scholes option pricing, Greeks and implied
##' volatility are also provided.
##'
##' Important functions include:
##' 
##' \code{\link{npv}}, \code{\link{irr}}, \code{\link{duration}}, 
##' \code{\link{annuity.pv}}, \code{\link{bond.price}}, \code{\link{bond.yield}},
##' \code{\link{GenBS}}, \code{\link{GenBSImplied}}
##' 
##' For more details, see the vignette
##' 
##' @name jrvFinance-package
##' @aliases jrvFinance-package jrvFinance
##' @docType package
##' 
##' @author Prof. Jayanth R. Varma \email{jrvarma@@iima.ac.in}
##' @references
##' The 30/360 day count was converted from C++ code in the QuantLib library.
##' The Newton Raphson solver was converted from C++ code in the Boost library

NULL





R/daycount.R
##' Day count and year fraction for bond pricing
##' 
##' Implements 30/360, ACT/360, ACT/360 and 30/360E day count conventions. 
##' 
##'
##' @rdname daycount
##' @name daycount
##' @aliases daycount.30.360 daycount.actual yearFraction
##' 
##' @param d1 The starting date of period for day counts
##' @param d2 The ending date of period for day counts
##' @param r1 The starting date of reference period for ACT/ACT day counts
##' @param r2 The ending date of reference period for ACT/ACT day counts
##' @param freq The frequency of coupon payments: 1 for annual, 2 for
##' semi-annual, 12 for monthly.
##' @param convention The daycount convention
##' @param variant Three variants of the 30/360 convention are implemented, but
##' only one variant of ACT/ACT is currently implemented
##' @author Prof. Jayanth R. Varma \email{jrvarma@@iima.ac.in}
##' @references The 30/360 day count was converted from C++ code in the
##' QuantLib library
NULL


##' @rdname daycount
##' @export
yearFraction <- function(d1, d2, r1, r2, freq=2,
                         convention=c("30/360", "ACT/ACT",
                                      "ACT/360", "30/360E")){
  convention <- match.arg(convention)
  if (convention == "ACT/ACT") (1/freq) * daycount.actual(d1,d2) /
                                   daycount.actual(r1,r2)
  else if (convention == "ACT/360") daycount.actual(d1,d2) / 360
  else if (convention == "30/360") daycount.30.360(d1,d2) / 360
  else if (convention == "30/360E") daycount.30.360(d1,d2, "E") / 360
}

##' @rdname daycount
##' @export
daycount.actual <- function(d1, d2, variant = c("bond")){
  as.integer(as.Date(d2) - as.Date(d1))
}

##' @rdname daycount
##' @export
daycount.30.360 <- function(d1, d2, variant = c("US", "EU", "IT")){
  ## The algorithm is taken from the QuantLib source code
  D1 <- as.POSIXlt(d1)
  D2 <- as.POSIXlt(d2)
  dd1 <- D1$mday
  dd2 <- D2$mday
  mm1 <- D1$mon
  mm2 <- D2$mon
  yy1 <- D1$year
  yy2 <- D2$year
  variant <- match.arg(variant)
  if (variant == "US" && dd2 == 31 && dd1 < 30) {
    dd2 <- 1
    mm2 <- mm2 + 1
  }
  if (variant == "IT" && mm1 == 2 && dd1 > 27) dd1 = 30
  if (variant == "IT" && mm2 == 2 && dd2 > 27) dd2 = 30
  360*(yy2-yy1) + 30*(mm2-mm1-1) + max(0,30-dd1) + min(30,dd2)
}

##' Shift date by a number of months
##' 
##' Convenience function for finding the same date in different months. Used for example
##' to find coupon dates of bonds given the maturity date. See \code{\link{coupons}}
##'
##' @param from starting date - a character string or any object that can be
##' converted into date using \code{\link{as.Date}}.
##' @param months Number of months (can be negative)
##' @export
edate <- function(from, months=1){
  from.lt <- as.POSIXlt(from)
  d <- from.lt$mday
  m <- from.lt$mon
  y <- from.lt$year + 1900
  M <- (m + months) %% 12
  Y <- y + (m + months - M) %/% 12
  ld <- last.day.of.month(m+1, y)
  LD <- last.day.of.month(M+1, Y)
  D <- ifelse (d == ld | d > LD, LD, d)
  as.Date(ISOdate(Y, M+1, D))
}

## Used internally not exported
last.day.of.month <- function(m, y){
  c(31,28,31, 30,31,30, 31,31,30, 31,30,31)[m] +
      ifelse(is.leap.year(y) & m == 2, 1, 0)
}

## Used internally not exported
is.leap.year <- function(y){
  y %% 4 == 0 & (y %% 100 !=0 | y %% 400 == 0)
}

